In [7]:
# Imports section
import pandas as pd

## Part 1. Loading the dataset

In [8]:
# Using pandas load the dataset (load remotely, not locally)
data = pd.read_csv('science_data_large.csv')
# Output the first 15 rows of the data
print(data.head(15))
# Display a summary of the table information (number of datapoints, etc.)
print(data.info())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## Part 2. Splitting the dataset

In [9]:
# Take the pandas dataset and split it into our features (X) and label (y)
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
from sklearn.model_selection import train_test_split

# Define features (X) and label (y)
X = data[['Temperature °C', 'Mols KCL']]  # Features
y = data['Size nm^3']  # Target label

# Split into training and testing sets (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

# Print the shape of training and test sets
print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)


X_train shape: (900, 2)
X_test shape: (100, 2)


## Part 3. Perform a Linear Regression

In [10]:
# Use sklearn to train a model on the training set

# Create a sample datapoint and predict the output of that sample with the trained model

# Report on the score for that model, in your own words (markdown, not code) explain what the score means

# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX

from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

# Initialize the Linear Regression model
model = LinearRegression()

# Train the model on the training data
model.fit(X_train, y_train)

# Make predictions on the test data
y_pred = model.predict(X_test)

# Calculate the R² score for the model
score = r2_score(y_test, y_pred)
print("R² Score:", score)

# Extract the coefficients
coefficients = model.coef_
print("Coefficients:", coefficients)

# Extract the intercept
intercept = model.intercept_
print("Intercept:", intercept)

R² Score: 0.8552472077276095
Coefficients: [ 866.14641337 1032.69506649]
Intercept: -409391.47958340764


In the context of the linear regression model shown in the code, the R² score (also called the coefficient of determination) is a measure of how well the model's predictions match the actual values. An R² score of 0.855 indicates that 85.5% of the variance in the dependent variable (y_test) is explained by the independent variables (X_test) in the model. In simpler terms, the R² score tells us how well the regression model fits the data. A score closer to 1 means the  model does a very good job at explaining the variability in the target variable, whereas a score closer to 0 would  indicate a poor fit. In this case, an R² of 0.855 suggests the model has a strong predictive ability for the given dataset.

Here's the Equation: h(x)=−409931.4795+866.1464⋅x1 +1032.6951⋅x2 

𝑥1 refers to Temperature (°C), which represents the heat exposure that the slime is subjected to in each experiment. This variable influences the slime’s growth based on the temperature levels provided in your dataset.

𝑥2 refers to Mols of KCl, which represents the amount of Potassium Chloride the slime is exposed to. This variable affects the chemical interaction with the slime and contributes to the changes in its size.

Sample equation: $E = mc^2$

## Part 4. Use Cross Validation

In [11]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data

# Report on their finding and their significance
# Get the coefficients and intercept
coefficients = model.coef_
intercept = model.intercept_

print("Coefficients:", coefficients)
print("Intercept:", intercept)


Coefficients: [ 866.14641337 1032.69506649]
Intercept: -409391.47958340764


Report on their finding and their significance

crross-validation helps us understand how well the model generalizes to unseen data. If the cross-validation scores show low variance (i.e., a small standard deviation), it means that the model’s performance is consistent across different splits of the data. If the scores have high variance, it suggests the model is sensitive to the specific data it was trained on.

The coefficients indicate that for every 1°C increase in temperature, the slime's size increases by approximately 866.15 nm³, and for every 1 mol increase in KCl, the size increases by about 1032.70 nm³. The intercept suggests the baseline value of slime size when both factors are zero, though this may not be physically meaningful. Cross-validation helps verify that the model generalizes well, ensuring consistent performance across different data splits without overfitting.



## Part 5. Using Polynomial Regression

In [12]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2

# Report on the metrics and output the resultant equation as you did in Part 3.

from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

# Assuming X_train, X_test, y_train, y_test are already defined

# Step 1: Create polynomial features
poly = PolynomialFeatures(degree=2)
X_train_poly = poly.fit_transform(X_train)
X_test_poly = poly.transform(X_test)

# Step 2: Train a new linear regression model on the polynomial features
poly_model = LinearRegression()
poly_model.fit(X_train_poly, y_train)

# Step 3: Make predictions on the test set
y_pred_poly = poly_model.predict(X_test_poly)

# Step 4: Calculate metrics for the polynomial regression
mse_poly = mean_squared_error(y_test, y_pred_poly)
r2_poly = r2_score(y_test, y_pred_poly)

# Output the results
print("Polynomial Regression (Degree 2) Results:")
print("Mean Squared Error:", mse_poly)
print("R-squared:", r2_poly)

# Step 5: Get the coefficients and intercept for the polynomial model
poly_coefficients = poly_model.coef_
poly_intercept = poly_model.intercept_

print("Coefficients:", poly_coefficients)
print("Intercept:", poly_intercept)

# Step 6: Output the polynomial equation (h(x))
# Step 6: Output the polynomial equation (h(x))
equation_terms = poly.get_feature_names_out()  # Automatically infer feature names
equation = " + ".join([f"{coeff:.4f}*{term}" for coeff, term in zip(poly_coefficients, equation_terms)])


Polynomial Regression (Degree 2) Results:
Mean Squared Error: 1.4693521677230201e-08
R-squared: 1.0
Coefficients: [ 0.00000000e+00  1.20000000e+01 -1.27195488e-07  1.26494371e-11
  2.00000000e+00  2.85714287e-02]
Intercept: 2.0477105863392353e-05


Regression Equation: h(x) = 12*a + 2*ab + 0.028571 * b^2

Metrics Report
the metrics for the polynomial regression model (degree 2) are:

Mean Squared Error (MSE): 1.469351677237021e−08 and 
R-squared: 1.0.
The extremely low Mean Squared Error suggests that the model's predictions are almost identical to the actual values. An R-squared value of 1.0 indicates a perfect fit, meaning the polynomial regression model accounts for 100% of the variation in the target variable. However, this raises concerns about possible overfitting. Overfitting occurs when a model captures not only the true patterns but also irrelevant noise in the data, leading to strong performance on the current dataset but potential difficulties in predicting new, unseen data. As a result, the model’s flawless performance here might not translate to good generalization, which can be misleading.

Polynomial Regression Equation
The extracted coefficients and intercept are used to form the polynomial regression equation. The equation for h(x) = 12*a + 2*ab + 0.028571 * b^2

This equation represents the relationship between the features (temperature, mols of KCl) and the target variable. Each term combines a feature with its corresponding coefficient, showcasing the interaction terms and squared terms that polynomial regression captures.

Summary
Polynomial regression model provides a perfect fit (R² = 1.0), meaning the model explains all the variability in the target variable based on the given features, leading to possible overfitting. The equation includes both interaction terms (e.g., Temperature °C Mols KCL and squared terms (e.g.,Mols KCL^2 ) which polynomial regression introduces to capture more complex relationships in the data.